src/array.rs

```diff
@@ -0,0 +1,456 @@
+        unsafe { mlx_sys::mlx_array_data_complex64(array.ptr) as *const Self }
+    }
+}
+
+pub struct Array {
+    ptr: mlx_array,
+}
+
+impl std::fmt::Debug for Array {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let description = crate::utils::mlx_describe(self.ptr as *mut c_void)
+            .unwrap_or_else(|| "Array".to_string());
+        write!(f, "{:?}", description)
+    }
+}
+
+// TODO: Clone should probably NOT be implemented because the underlying pointer is atomically
+// reference counted but not guarded by a mutex.
+
+impl Drop for Array {
+    fn drop(&mut self) {
+        // TODO: check memory leak with some tool?
+
+        // Decrease the reference count
+        unsafe { mlx_sys::mlx_free(self.ptr as *mut c_void) };
+    }
+}
+
+impl Array {
+    /// Create a new array from an existing mlx_array pointer.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure the reference count of the array is properly incremented with
+    /// `mlx_sys::mlx_retain`.
+    pub unsafe fn from_ptr(ptr: mlx_array) -> Array {
+        Array { ptr }
+    }
+
+    // TODO: should this be unsafe?
+    pub fn as_ptr(&self) -> mlx_array {
+        self.ptr
+    }
+
+    /// New array from a bool scalar.
+    pub fn from_bool(val: bool) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_bool(val) };
+        Array { ptr }
+    }
+
+    /// New array from a int scalar.
+    pub fn from_int(val: i32) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_int(val) };
+        Array { ptr }
+    }
+
+    /// New array from a float scalar.
+    pub fn from_float(val: f32) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_float(val) };
+        Array { ptr }
+    }
+
+    /// New array from a complex scalar.
+    pub fn from_complex(val: c64) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_complex(val.re, val.im) };
+        Array { ptr }
+    }
+
+    /// New array from existing buffer.
+    ///
+    /// # Parameters
+    ///
+    /// - `data`: A buffer which will be copied.
+    /// - `shape`: Shape of the array.
+    /// - `dim`: Number of dimensions (size of shape).
+    pub fn from_data<T: ArrayElement>(data: &[T], shape: &[i32], dim: i32) -> Self {
+        let ptr = unsafe {
+            mlx_sys::mlx_array_from_data(
+                data.as_ptr() as *const c_void,
+                shape.as_ptr(),
+                dim,
+                T::DTYPE as u32,
+            )
+        };
+
+        Array { ptr }
+    }
+
+    /// The size of the array’s datatype in bytes.
+    pub fn itemsize(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_itemsize(self.ptr) }
+    }
+
+    /// Number of elements in the array.
+    pub fn size(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_size(self.ptr) }
+    }
+
+    /// The strides of the array.
+    pub fn strides(&self) -> &[usize] {
+        let ndim = self.ndim();
```

Comment:
I think this is expected behavior. `mlx` follows the `numpy` api, and `strides` and `shape` of a scalar array in `mlx` and `numpy` would be empty.

```python
import mlx.core as mx
import numpy as np

a = mx.array(1)
# mlx python doesn't expose strides somehow
a.shape # would give ()

b = np.array(1)
b.strides # would give ()
b.shape # would give ()
```

Reply:
Same thing found in the swift binding. Found the code below in the [swift tutorial](https://github.com/ml-explore/mlx-swift/blob/main/Source/Examples/Tutorial.swift)

```swift
// scalars have 0 dimensions
let ndim = x.ndim
assert(ndim == 0)

// scalar shapes are empty arrays
let shape = x.shape
assert(shape == [])
```

Reply:
```swift
    public var strides: [Int] {
        let ndim = mlx_array_ndim(ctx)
        guard ndim > 0 else { return [] }
        let strides = mlx_array_strides(ctx)!
        return (0 ..< ndim).map { Int(strides[$0]) }
    }
```

Reply:
> we should add a check here to make sure ndim is > 0 otherwise let's return empty.

I misread. This is already returning an empty slice if `ndim == 0`. Check the scalar array unit test